xformers/ops/memory_efficient_attention.py

```diff
@@ -722,6 +733,135 @@ def _flash_attn_backward(
         return dq, dk, dv, softmax_d
 
 
+class _FakeContext(SimpleNamespace):
+    def save_for_backward(self, *args):
+        return
+
+
+class TritonFlashAttentionOp(AttentionOpBase):
+    FORWARD_OPERATOR = None
+    SUPPORTED_DEVICES = {"cuda"}
+    SUPPORTED_DTYPES = {torch.half, torch.bfloat16}
+    SUPPORTED_MAX_K: float = 128
+    SUPPORTED_ATTN_BIAS_TYPES: Set[Any] = {
+        type(None),
+        LowerTriangularMask,
+        # TODO: backwards accuracy is failing for a few cases, perhaps we want to disable this for now.
+        # torch.Tensor,
+    }
+    SUPPORTS_DROPOUT = False
+    SUPPORTS_CUSTOM_SCALE = True
+    NAME = "tritonflashatt"
+
+    @classmethod
+    def info(cls):
+        if not has_triton_flashattention:
+            return "not built"
+        return "available"
+
+    @classmethod
+    def supports(cls, d: "AttentionOpDispatch") -> bool:
+        if not has_triton_flashattention:
+            return False
+        device_capability = torch.cuda.get_device_capability(d.device)
+        is_sm80 = device_capability[0] >= 8
+        if not is_sm80:
+            return False
+        return super(TritonFlashAttentionOp, cls).supports(d)
+
+    @classmethod
+    def forward_no_grad(
+        cls,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_bias: Optional[Union[torch.Tensor, AttentionMask]],
+        p: float,
+        scale: Optional[float] = None,
+    ) -> torch.Tensor:
+        ctx = _FakeContext()
+        return cls.forward(
+            ctx=ctx,
+            query=query,
+            key=key,
+            value=value,
+            attn_bias=attn_bias,
+            p=p,
+            scale=scale,
+        )
+
+    @classmethod
+    def forward(cls, ctx, query, key, value, attn_bias, p, scale):
+        softmax_scale = query.shape[-1] ** (-0.5) if scale is None else scale
+        causal = isinstance(attn_bias, LowerTriangularMask)
+        if not causal and attn_bias is not None and attn_bias.ndim == 3:
+            B = query.shape[0]
+            h = attn_bias.shape[0] // B
+            attn_bias = attn_bias.reshape(B, h, attn_bias.shape[1], attn_bias.shape[2])
+
+        return triton_flash.forward(
```

Comment:
I would have probably used the functional implementations in here [(like `_flash_attn_forward` and `_flash_attn_backward`)](https://github.com/HazyResearch/flash-attention/blob/0bf5e50038ee341ece03bfd0c8ff45a6c57aed5a/flash_attn/flash_attn_triton.py#L581). Although they are private functions, we have flashattention as a submodule and thus we control the version being used. And this means we don't need to do `_FakeContext` workarounds (which I'm not sure to what extent they are robust)

Reply:
Thanks for the suggestion, I've updated it to use `_flash_attn_forward` and `_flash_attn_backward` and removed the `_FakeContext`